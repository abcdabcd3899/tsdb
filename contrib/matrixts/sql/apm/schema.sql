-- start_ignore
CREATE EXTENSION matrixts;
SET enable_mergejoin TO off;
SET enable_nestloop TO off;
SET enable_parallel_mode TO off;
SET gp_enable_multiphase_agg TO off;
SET optimizer TO off;
CREATE SCHEMA IF NOT EXISTS apm_test;
SET matrix.ts_guc_apm_allow_dml TO ON;
TRUNCATE matrixts_internal.apm_operation_log;
TRUNCATE matrixts_internal.apm_rel_policy;
TRUNCATE matrixts_internal.apm_rel_policy_action;
SET matrix.ts_guc_apm_allow_dml TO OFF;
-- end_ignore

---------------------------------------------
-- List function
---- apm_generic_list_partitions
---------------------------------------------

-- start_ignore
DROP TABLE IF EXISTS apm_test.sales;
CREATE TABLE apm_test.sales (id int, year int, qtr int, day int, region text)
DISTRIBUTED BY (id)
PARTITION BY RANGE (year)
(PARTITION yr START (2010) END (2014) EVERY (1));
-- end_ignore

SELECT
    root_reloid IS NULL AS root_reloid
  , root_relname
  , parent_reloid IS NULL AS parent_reloid
  , parent_relname
  , reloid IS NULL AS reloid
  , relname
  , relnsp
  , partitiontype
  , partitionlevel
  , partitionisdefault
  , partitionboundary
  , addopt
FROM matrixts_internal.apm_generic_list_partitions('apm_test.sales'::regclass, NULL)
ORDER BY relname;

-- start_ignore
DROP TABLE IF EXISTS apm_test.sales2;
CREATE TABLE apm_test.sales2 (trans_id int, date date, amount
decimal(9,2), region text)
DISTRIBUTED BY (trans_id)
PARTITION BY RANGE (date)
SUBPARTITION BY LIST (region)
SUBPARTITION TEMPLATE
( SUBPARTITION usa VALUES ('usa'),
  SUBPARTITION asia VALUES ('asia'),
  SUBPARTITION europe VALUES ('europe'),
  DEFAULT SUBPARTITION other_regions)
  (START (date '2011-01-01') INCLUSIVE
   END (date '2012-01-01') EXCLUSIVE
   EVERY (INTERVAL '1 month'),
   DEFAULT PARTITION outlying_dates);
-- end_ignore

SELECT
    root_reloid IS NULL AS root_reloid
  , root_relname
  , parent_reloid IS NULL AS parent_reloid
  , parent_relname
  , reloid IS NULL AS reloid
  , relname
  , relnsp
  , partitiontype
  , partitionlevel
  , partitionisdefault
  , partitionboundary
  , addopt
FROM matrixts_internal.apm_generic_list_partitions('apm_test.sales2'::regclass, NULL)
ORDER BY relname;

---------------------------------------------
-- Schema protection
---------------------------------------------

-- apm_schema_protect event trigger should not crash normal DDL
CREATE TABLE apm_test.to_alter(id int, id2 int);
ALTER TABLE apm_test.to_alter SET DISTRIBUTED BY (id2);
SET search_path TO apm_test;
ALTER TABLE to_alter SET DISTRIBUTED BY (id);
RESET search_path;

-- should ban DDL changes
ALTER TYPE matrixts_internal.apm_partition_item ADD ATTRIBUTE test int;
ALTER TYPE matrixts_internal.apm_partition_item DROP ATTRIBUTE test;

ALTER TABLE matrixts_internal.apm_policy_class ADD COLUMN test int;
ALTER TABLE matrixts_internal.apm_policy_class DROP COLUMN test;

ALTER TABLE matrixts_internal.apm_policy_action ADD COLUMN test int;
ALTER TABLE matrixts_internal.apm_policy_action DROP COLUMN test;

ALTER TABLE matrixts_internal.apm_rel_policy ADD COLUMN test int;
ALTER TABLE matrixts_internal.apm_rel_policy DROP COLUMN test;

ALTER TABLE matrixts_internal.apm_rel_policy_action ADD COLUMN test int;
ALTER TABLE matrixts_internal.apm_rel_policy_action DROP COLUMN test;

SET search_path TO matrixts_internal;
ALTER TABLE apm_rel_policy_action ADD COLUMN extra text;
ALTER TABLE apm_rel_policy_action SET DISTRIBUTED BY (action);
RESET search_path;

-- should allow DDL changes with internal GUC set
SET matrix.ts_guc_apm_allow_ddl TO ON;

ALTER TYPE matrixts_internal.apm_partition_item ADD ATTRIBUTE test int;
ALTER TYPE matrixts_internal.apm_partition_item DROP ATTRIBUTE test;

ALTER TABLE matrixts_internal.apm_policy_class ADD COLUMN test int;
ALTER TABLE matrixts_internal.apm_policy_class DROP COLUMN test;

ALTER TABLE matrixts_internal.apm_policy_action ADD COLUMN test int;
ALTER TABLE matrixts_internal.apm_policy_action DROP COLUMN test;

ALTER TABLE matrixts_internal.apm_rel_policy ADD COLUMN test int;
ALTER TABLE matrixts_internal.apm_rel_policy DROP COLUMN test;

ALTER TABLE matrixts_internal.apm_rel_policy_action ADD COLUMN test int;
ALTER TABLE matrixts_internal.apm_rel_policy_action DROP COLUMN test;

RESET matrix.ts_guc_apm_allow_ddl;

-- should say "table has already been expanded"
-- not ERROR:  MatrixTS: cannot alter metatable apm_action_log
ALTER TABLE matrixts_internal.apm_action_log EXPAND TABLE;
