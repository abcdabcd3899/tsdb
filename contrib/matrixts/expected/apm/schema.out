-- start_ignore
CREATE EXTENSION matrixts;
SET enable_mergejoin TO off;
SET enable_nestloop TO off;
SET enable_parallel_mode TO off;
SET gp_enable_multiphase_agg TO off;
SET optimizer TO off;
CREATE SCHEMA IF NOT EXISTS apm_test;
SET matrix.ts_guc_apm_allow_dml TO ON;
TRUNCATE matrixts_internal.apm_operation_log;
TRUNCATE matrixts_internal.apm_rel_policy;
TRUNCATE matrixts_internal.apm_rel_policy_action;
SET matrix.ts_guc_apm_allow_dml TO OFF;
-- end_ignore
---------------------------------------------
-- List function
---- apm_generic_list_partitions
---------------------------------------------
-- start_ignore
DROP TABLE IF EXISTS apm_test.sales;
NOTICE:  table "sales" does not exist, skipping
CREATE TABLE apm_test.sales (id int, year int, qtr int, day int, region text)
DISTRIBUTED BY (id)
PARTITION BY RANGE (year)
(PARTITION yr START (2010) END (2014) EVERY (1));
-- end_ignore
SELECT
    root_reloid IS NULL AS root_reloid
  , root_relname
  , parent_reloid IS NULL AS parent_reloid
  , parent_relname
  , reloid IS NULL AS reloid
  , relname
  , relnsp
  , partitiontype
  , partitionlevel
  , partitionisdefault
  , partitionboundary
  , addopt
FROM matrixts_internal.apm_generic_list_partitions('apm_test.sales'::regclass, NULL)
ORDER BY relname;
 root_reloid | root_relname | parent_reloid | parent_relname | reloid |     relname      |  relnsp  | partitiontype | partitionlevel | partitionisdefault |        partitionboundary         | addopt 
-------------+--------------+---------------+----------------+--------+------------------+----------+---------------+----------------+--------------------+----------------------------------+--------
 f           | sales        | t             |                | f      | sales            | apm_test |               |              0 | f                  |                                  | 
 f           | sales        | f             | sales          | f      | sales_1_prt_yr_1 | apm_test | range         |              1 | f                  | FOR VALUES FROM (2010) TO (2011) | 
 f           | sales        | f             | sales          | f      | sales_1_prt_yr_2 | apm_test | range         |              1 | f                  | FOR VALUES FROM (2011) TO (2012) | 
 f           | sales        | f             | sales          | f      | sales_1_prt_yr_3 | apm_test | range         |              1 | f                  | FOR VALUES FROM (2012) TO (2013) | 
 f           | sales        | f             | sales          | f      | sales_1_prt_yr_4 | apm_test | range         |              1 | f                  | FOR VALUES FROM (2013) TO (2014) | 
(5 rows)

-- start_ignore
DROP TABLE IF EXISTS apm_test.sales2;
NOTICE:  table "sales2" does not exist, skipping
CREATE TABLE apm_test.sales2 (trans_id int, date date, amount
decimal(9,2), region text)
DISTRIBUTED BY (trans_id)
PARTITION BY RANGE (date)
SUBPARTITION BY LIST (region)
SUBPARTITION TEMPLATE
( SUBPARTITION usa VALUES ('usa'),
  SUBPARTITION asia VALUES ('asia'),
  SUBPARTITION europe VALUES ('europe'),
  DEFAULT SUBPARTITION other_regions)
  (START (date '2011-01-01') INCLUSIVE
   END (date '2012-01-01') EXCLUSIVE
   EVERY (INTERVAL '1 month'),
   DEFAULT PARTITION outlying_dates);
-- end_ignore
SELECT
    root_reloid IS NULL AS root_reloid
  , root_relname
  , parent_reloid IS NULL AS parent_reloid
  , parent_relname
  , reloid IS NULL AS reloid
  , relname
  , relnsp
  , partitiontype
  , partitionlevel
  , partitionisdefault
  , partitionboundary
  , addopt
FROM matrixts_internal.apm_generic_list_partitions('apm_test.sales2'::regclass, NULL)
ORDER BY relname;
 root_reloid | root_relname | parent_reloid |       parent_relname        | reloid |                     relname                     |  relnsp  | partitiontype | partitionlevel | partitionisdefault |                partitionboundary                 | addopt 
-------------+--------------+---------------+-----------------------------+--------+-------------------------------------------------+----------+---------------+----------------+--------------------+--------------------------------------------------+--------
 f           | sales2       | t             |                             | f      | sales2                                          | apm_test |               |              0 | f                  |                                                  | 
 f           | sales2       | f             | sales2                      | f      | sales2_1_prt_10                                 | apm_test | range         |              1 | f                  | FOR VALUES FROM ('09-01-2011') TO ('10-01-2011') | 
 f           | sales2       | f             | sales2_1_prt_10             | f      | sales2_1_prt_10_2_prt_asia                      | apm_test | list          |              2 | f                  | FOR VALUES IN ('asia')                           | 
 f           | sales2       | f             | sales2_1_prt_10             | f      | sales2_1_prt_10_2_prt_europe                    | apm_test | list          |              2 | f                  | FOR VALUES IN ('europe')                         | 
 f           | sales2       | f             | sales2_1_prt_10             | f      | sales2_1_prt_10_2_prt_other_regions             | apm_test | list          |              2 | t                  | DEFAULT                                          | 
 f           | sales2       | f             | sales2_1_prt_10             | f      | sales2_1_prt_10_2_prt_usa                       | apm_test | list          |              2 | f                  | FOR VALUES IN ('usa')                            | 
 f           | sales2       | f             | sales2                      | f      | sales2_1_prt_11                                 | apm_test | range         |              1 | f                  | FOR VALUES FROM ('10-01-2011') TO ('11-01-2011') | 
 f           | sales2       | f             | sales2_1_prt_11             | f      | sales2_1_prt_11_2_prt_asia                      | apm_test | list          |              2 | f                  | FOR VALUES IN ('asia')                           | 
 f           | sales2       | f             | sales2_1_prt_11             | f      | sales2_1_prt_11_2_prt_europe                    | apm_test | list          |              2 | f                  | FOR VALUES IN ('europe')                         | 
 f           | sales2       | f             | sales2_1_prt_11             | f      | sales2_1_prt_11_2_prt_other_regions             | apm_test | list          |              2 | t                  | DEFAULT                                          | 
 f           | sales2       | f             | sales2_1_prt_11             | f      | sales2_1_prt_11_2_prt_usa                       | apm_test | list          |              2 | f                  | FOR VALUES IN ('usa')                            | 
 f           | sales2       | f             | sales2                      | f      | sales2_1_prt_12                                 | apm_test | range         |              1 | f                  | FOR VALUES FROM ('11-01-2011') TO ('12-01-2011') | 
 f           | sales2       | f             | sales2_1_prt_12             | f      | sales2_1_prt_12_2_prt_asia                      | apm_test | list          |              2 | f                  | FOR VALUES IN ('asia')                           | 
 f           | sales2       | f             | sales2_1_prt_12             | f      | sales2_1_prt_12_2_prt_europe                    | apm_test | list          |              2 | f                  | FOR VALUES IN ('europe')                         | 
 f           | sales2       | f             | sales2_1_prt_12             | f      | sales2_1_prt_12_2_prt_other_regions             | apm_test | list          |              2 | t                  | DEFAULT                                          | 
 f           | sales2       | f             | sales2_1_prt_12             | f      | sales2_1_prt_12_2_prt_usa                       | apm_test | list          |              2 | f                  | FOR VALUES IN ('usa')                            | 
 f           | sales2       | f             | sales2                      | f      | sales2_1_prt_13                                 | apm_test | range         |              1 | f                  | FOR VALUES FROM ('12-01-2011') TO ('01-01-2012') | 
 f           | sales2       | f             | sales2_1_prt_13             | f      | sales2_1_prt_13_2_prt_asia                      | apm_test | list          |              2 | f                  | FOR VALUES IN ('asia')                           | 
 f           | sales2       | f             | sales2_1_prt_13             | f      | sales2_1_prt_13_2_prt_europe                    | apm_test | list          |              2 | f                  | FOR VALUES IN ('europe')                         | 
 f           | sales2       | f             | sales2_1_prt_13             | f      | sales2_1_prt_13_2_prt_other_regions             | apm_test | list          |              2 | t                  | DEFAULT                                          | 
 f           | sales2       | f             | sales2_1_prt_13             | f      | sales2_1_prt_13_2_prt_usa                       | apm_test | list          |              2 | f                  | FOR VALUES IN ('usa')                            | 
 f           | sales2       | f             | sales2                      | f      | sales2_1_prt_2                                  | apm_test | range         |              1 | f                  | FOR VALUES FROM ('01-01-2011') TO ('02-01-2011') | 
 f           | sales2       | f             | sales2_1_prt_2              | f      | sales2_1_prt_2_2_prt_asia                       | apm_test | list          |              2 | f                  | FOR VALUES IN ('asia')                           | 
 f           | sales2       | f             | sales2_1_prt_2              | f      | sales2_1_prt_2_2_prt_europe                     | apm_test | list          |              2 | f                  | FOR VALUES IN ('europe')                         | 
 f           | sales2       | f             | sales2_1_prt_2              | f      | sales2_1_prt_2_2_prt_other_regions              | apm_test | list          |              2 | t                  | DEFAULT                                          | 
 f           | sales2       | f             | sales2_1_prt_2              | f      | sales2_1_prt_2_2_prt_usa                        | apm_test | list          |              2 | f                  | FOR VALUES IN ('usa')                            | 
 f           | sales2       | f             | sales2                      | f      | sales2_1_prt_3                                  | apm_test | range         |              1 | f                  | FOR VALUES FROM ('02-01-2011') TO ('03-01-2011') | 
 f           | sales2       | f             | sales2_1_prt_3              | f      | sales2_1_prt_3_2_prt_asia                       | apm_test | list          |              2 | f                  | FOR VALUES IN ('asia')                           | 
 f           | sales2       | f             | sales2_1_prt_3              | f      | sales2_1_prt_3_2_prt_europe                     | apm_test | list          |              2 | f                  | FOR VALUES IN ('europe')                         | 
 f           | sales2       | f             | sales2_1_prt_3              | f      | sales2_1_prt_3_2_prt_other_regions              | apm_test | list          |              2 | t                  | DEFAULT                                          | 
 f           | sales2       | f             | sales2_1_prt_3              | f      | sales2_1_prt_3_2_prt_usa                        | apm_test | list          |              2 | f                  | FOR VALUES IN ('usa')                            | 
 f           | sales2       | f             | sales2                      | f      | sales2_1_prt_4                                  | apm_test | range         |              1 | f                  | FOR VALUES FROM ('03-01-2011') TO ('04-01-2011') | 
 f           | sales2       | f             | sales2_1_prt_4              | f      | sales2_1_prt_4_2_prt_asia                       | apm_test | list          |              2 | f                  | FOR VALUES IN ('asia')                           | 
 f           | sales2       | f             | sales2_1_prt_4              | f      | sales2_1_prt_4_2_prt_europe                     | apm_test | list          |              2 | f                  | FOR VALUES IN ('europe')                         | 
 f           | sales2       | f             | sales2_1_prt_4              | f      | sales2_1_prt_4_2_prt_other_regions              | apm_test | list          |              2 | t                  | DEFAULT                                          | 
 f           | sales2       | f             | sales2_1_prt_4              | f      | sales2_1_prt_4_2_prt_usa                        | apm_test | list          |              2 | f                  | FOR VALUES IN ('usa')                            | 
 f           | sales2       | f             | sales2                      | f      | sales2_1_prt_5                                  | apm_test | range         |              1 | f                  | FOR VALUES FROM ('04-01-2011') TO ('05-01-2011') | 
 f           | sales2       | f             | sales2_1_prt_5              | f      | sales2_1_prt_5_2_prt_asia                       | apm_test | list          |              2 | f                  | FOR VALUES IN ('asia')                           | 
 f           | sales2       | f             | sales2_1_prt_5              | f      | sales2_1_prt_5_2_prt_europe                     | apm_test | list          |              2 | f                  | FOR VALUES IN ('europe')                         | 
 f           | sales2       | f             | sales2_1_prt_5              | f      | sales2_1_prt_5_2_prt_other_regions              | apm_test | list          |              2 | t                  | DEFAULT                                          | 
 f           | sales2       | f             | sales2_1_prt_5              | f      | sales2_1_prt_5_2_prt_usa                        | apm_test | list          |              2 | f                  | FOR VALUES IN ('usa')                            | 
 f           | sales2       | f             | sales2                      | f      | sales2_1_prt_6                                  | apm_test | range         |              1 | f                  | FOR VALUES FROM ('05-01-2011') TO ('06-01-2011') | 
 f           | sales2       | f             | sales2_1_prt_6              | f      | sales2_1_prt_6_2_prt_asia                       | apm_test | list          |              2 | f                  | FOR VALUES IN ('asia')                           | 
 f           | sales2       | f             | sales2_1_prt_6              | f      | sales2_1_prt_6_2_prt_europe                     | apm_test | list          |              2 | f                  | FOR VALUES IN ('europe')                         | 
 f           | sales2       | f             | sales2_1_prt_6              | f      | sales2_1_prt_6_2_prt_other_regions              | apm_test | list          |              2 | t                  | DEFAULT                                          | 
 f           | sales2       | f             | sales2_1_prt_6              | f      | sales2_1_prt_6_2_prt_usa                        | apm_test | list          |              2 | f                  | FOR VALUES IN ('usa')                            | 
 f           | sales2       | f             | sales2                      | f      | sales2_1_prt_7                                  | apm_test | range         |              1 | f                  | FOR VALUES FROM ('06-01-2011') TO ('07-01-2011') | 
 f           | sales2       | f             | sales2_1_prt_7              | f      | sales2_1_prt_7_2_prt_asia                       | apm_test | list          |              2 | f                  | FOR VALUES IN ('asia')                           | 
 f           | sales2       | f             | sales2_1_prt_7              | f      | sales2_1_prt_7_2_prt_europe                     | apm_test | list          |              2 | f                  | FOR VALUES IN ('europe')                         | 
 f           | sales2       | f             | sales2_1_prt_7              | f      | sales2_1_prt_7_2_prt_other_regions              | apm_test | list          |              2 | t                  | DEFAULT                                          | 
 f           | sales2       | f             | sales2_1_prt_7              | f      | sales2_1_prt_7_2_prt_usa                        | apm_test | list          |              2 | f                  | FOR VALUES IN ('usa')                            | 
 f           | sales2       | f             | sales2                      | f      | sales2_1_prt_8                                  | apm_test | range         |              1 | f                  | FOR VALUES FROM ('07-01-2011') TO ('08-01-2011') | 
 f           | sales2       | f             | sales2_1_prt_8              | f      | sales2_1_prt_8_2_prt_asia                       | apm_test | list          |              2 | f                  | FOR VALUES IN ('asia')                           | 
 f           | sales2       | f             | sales2_1_prt_8              | f      | sales2_1_prt_8_2_prt_europe                     | apm_test | list          |              2 | f                  | FOR VALUES IN ('europe')                         | 
 f           | sales2       | f             | sales2_1_prt_8              | f      | sales2_1_prt_8_2_prt_other_regions              | apm_test | list          |              2 | t                  | DEFAULT                                          | 
 f           | sales2       | f             | sales2_1_prt_8              | f      | sales2_1_prt_8_2_prt_usa                        | apm_test | list          |              2 | f                  | FOR VALUES IN ('usa')                            | 
 f           | sales2       | f             | sales2                      | f      | sales2_1_prt_9                                  | apm_test | range         |              1 | f                  | FOR VALUES FROM ('08-01-2011') TO ('09-01-2011') | 
 f           | sales2       | f             | sales2_1_prt_9              | f      | sales2_1_prt_9_2_prt_asia                       | apm_test | list          |              2 | f                  | FOR VALUES IN ('asia')                           | 
 f           | sales2       | f             | sales2_1_prt_9              | f      | sales2_1_prt_9_2_prt_europe                     | apm_test | list          |              2 | f                  | FOR VALUES IN ('europe')                         | 
 f           | sales2       | f             | sales2_1_prt_9              | f      | sales2_1_prt_9_2_prt_other_regions              | apm_test | list          |              2 | t                  | DEFAULT                                          | 
 f           | sales2       | f             | sales2_1_prt_9              | f      | sales2_1_prt_9_2_prt_usa                        | apm_test | list          |              2 | f                  | FOR VALUES IN ('usa')                            | 
 f           | sales2       | f             | sales2                      | f      | sales2_1_prt_outlying_dates                     | apm_test | range         |              1 | t                  | DEFAULT                                          | 
 f           | sales2       | f             | sales2_1_prt_outlying_dates | f      | sales2_1_prt_outlying_dates_2_prt_asia          | apm_test | list          |              2 | f                  | FOR VALUES IN ('asia')                           | 
 f           | sales2       | f             | sales2_1_prt_outlying_dates | f      | sales2_1_prt_outlying_dates_2_prt_europe        | apm_test | list          |              2 | f                  | FOR VALUES IN ('europe')                         | 
 f           | sales2       | f             | sales2_1_prt_outlying_dates | f      | sales2_1_prt_outlying_dates_2_prt_other_regions | apm_test | list          |              2 | t                  | DEFAULT                                          | 
 f           | sales2       | f             | sales2_1_prt_outlying_dates | f      | sales2_1_prt_outlying_dates_2_prt_usa           | apm_test | list          |              2 | f                  | FOR VALUES IN ('usa')                            | 
(66 rows)

---------------------------------------------
-- Schema protection
---------------------------------------------
-- apm_schema_protect event trigger should not crash normal DDL
CREATE TABLE apm_test.to_alter(id int, id2 int);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'id' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
ALTER TABLE apm_test.to_alter SET DISTRIBUTED BY (id2);
SET search_path TO apm_test;
ALTER TABLE to_alter SET DISTRIBUTED BY (id);
RESET search_path;
-- should ban DDL changes
ALTER TYPE matrixts_internal.apm_partition_item ADD ATTRIBUTE test int;
ERROR:  MatrixTS: cannot alter metatable apm_partition_item
ALTER TYPE matrixts_internal.apm_partition_item DROP ATTRIBUTE test;
ERROR:  MatrixTS: cannot alter metatable apm_partition_item
ALTER TABLE matrixts_internal.apm_policy_class ADD COLUMN test int;
ERROR:  MatrixTS: cannot alter metatable apm_policy_class
ALTER TABLE matrixts_internal.apm_policy_class DROP COLUMN test;
ERROR:  MatrixTS: cannot alter metatable apm_policy_class
ALTER TABLE matrixts_internal.apm_policy_action ADD COLUMN test int;
ERROR:  MatrixTS: cannot alter metatable apm_policy_action
ALTER TABLE matrixts_internal.apm_policy_action DROP COLUMN test;
ERROR:  MatrixTS: cannot alter metatable apm_policy_action
ALTER TABLE matrixts_internal.apm_rel_policy ADD COLUMN test int;
ERROR:  MatrixTS: cannot alter metatable apm_rel_policy
ALTER TABLE matrixts_internal.apm_rel_policy DROP COLUMN test;
ERROR:  MatrixTS: cannot alter metatable apm_rel_policy
ALTER TABLE matrixts_internal.apm_rel_policy_action ADD COLUMN test int;
ERROR:  MatrixTS: cannot alter metatable apm_rel_policy_action
ALTER TABLE matrixts_internal.apm_rel_policy_action DROP COLUMN test;
ERROR:  MatrixTS: cannot alter metatable apm_rel_policy_action
SET search_path TO matrixts_internal;
ALTER TABLE apm_rel_policy_action ADD COLUMN extra text;
ERROR:  MatrixTS: cannot alter metatable apm_rel_policy_action
ALTER TABLE apm_rel_policy_action SET DISTRIBUTED BY (action);
ERROR:  MatrixTS: cannot alter metatable apm_rel_policy_action
RESET search_path;
-- should allow DDL changes with internal GUC set
SET matrix.ts_guc_apm_allow_ddl TO ON;
ALTER TYPE matrixts_internal.apm_partition_item ADD ATTRIBUTE test int;
ALTER TYPE matrixts_internal.apm_partition_item DROP ATTRIBUTE test;
ALTER TABLE matrixts_internal.apm_policy_class ADD COLUMN test int;
ALTER TABLE matrixts_internal.apm_policy_class DROP COLUMN test;
ALTER TABLE matrixts_internal.apm_policy_action ADD COLUMN test int;
ALTER TABLE matrixts_internal.apm_policy_action DROP COLUMN test;
ALTER TABLE matrixts_internal.apm_rel_policy ADD COLUMN test int;
ALTER TABLE matrixts_internal.apm_rel_policy DROP COLUMN test;
ALTER TABLE matrixts_internal.apm_rel_policy_action ADD COLUMN test int;
ALTER TABLE matrixts_internal.apm_rel_policy_action DROP COLUMN test;
RESET matrix.ts_guc_apm_allow_ddl;
-- should say "table has already been expanded"
-- not ERROR:  MatrixTS: cannot alter metatable apm_action_log
ALTER TABLE matrixts_internal.apm_action_log EXPAND TABLE;
ERROR:  cannot expand table "apm_action_log"
DETAIL:  table has already been expanded
